File: javascript/packages/core/src/models/repairOrder.js
/** The central aggregate. */

import { withReference } from "../concerns/hasReference.js";
import { IllegalTransitionError } from "../errors.js";
import { Money } from "../money.js";
import { RepairStatus } from "../support/status.js";

/**
 * One part line on an order.
 *
 * @typedef {object} OrderPart
 * @property {number} partId
 * @property {string} sku
 * @property {number} quantity
 * @property {Money} unitPrice
 */

/**
 * Append-only lifecycle entry.
 *
 * @typedef {object} StatusChange
 * @property {import("../support/status.js").RepairStatusValue} from
 * @property {import("../support/status.js").RepairStatusValue} to
 * @property {string} changedBy
 */

/** A repair job: device in, invoice out. */
export class RepairOrder {
    /**
     * @param {number} id
     * @param {number} customerId
     * @param {number} deviceId
     */
    constructor(id, customerId, deviceId) {
        this.id = id;
        this.customerId = customerId;
        this.deviceId = deviceId;

        /** @type {import("../support/status.js").RepairStatusValue} */
        this.status = RepairStatus.Received;

        /** @type {import("../support/priority.js").PriorityValue} */
        this.priority = "standard";

        this.laborMinutes = 0;

        /** @type {OrderPart[]} */
        this.parts = [];

        /** @type {StatusChange[]} */
        this.log = [];
    }

    /**
     * Move to `next` when the lifecycle allows it, logging the change.
     *
     * @param {import("../support/status.js").RepairStatusValue} next
     * @param {string} changedBy
     * @returns {boolean}
     */
    transitionTo(next, changedBy) {
        if (!RepairStatus.transitionsTo(this.status).includes(next)) {
            return false;
        }

        this.log.push({ from: this.status, to: next, changedBy });
        this.status = next;

        return true;
    }

    /**
     * Drive the order to Completed, throwing on an illegal jump.
     *
     * @param {string} changedBy
     * @returns {void}
     */
    complete(changedBy) {
        if (!this.transitionTo(RepairStatus.Completed, changedBy)) {
            throw new IllegalTransitionError(this.status, RepairStatus.Completed);
        }
    }

    /**
     * Parts-only subtotal; labour is the calculator's business.
     *
     * @returns {Money}
     */
    partsSubtotal() {
        return Money.sum(this.parts.map((line) => line.unitPrice.times(line.quantity)));
    }

    /**
     * Total through the bound strategy: the concrete class is the container's choice.
     *
     * @param {import("../container.js").Container} container
     * @returns {Money}
     */
    total(container) {
        return container.invoiceCalculator().calculate(this);
    }

    /**
     * Attach a part line, priced from the stock record.
     *
     * @param {import("./part.js").Part} part
     * @param {number} quantity
     * @returns {void}
     */
    addPart(part, quantity) {
        this.parts.push({ partId: part.id, sku: part.sku, quantity, unitPrice: part.unitPrice });
    }

    /** @returns {boolean} */
    isOpen() {
        return RepairStatus.isOpen(this.status);
    }

    /**
     * @param {number} id
     * @param {number} customerId
     * @param {number} deviceId
     * @returns {RepairOrder}
     */
    static seed(id, customerId, deviceId) {
        const order = new RepairOrder(id, customerId, deviceId);
        order.referenceNumber = id;

        return order;
    }
}

withReference(RepairOrder);

File: rust/crates/atelier-core/src/models/repair_order.rs
//! The central aggregate.

use crate::concerns::has_reference::HasReference;
use crate::container::Container;
use crate::errors::AtelierError;
use crate::models::part::Part;
use crate::money::Money;
use crate::support::priority::Priority;
use crate::support::status::RepairStatus;
use serde::{Deserialize, Serialize};

/// One part line on an order.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct OrderPart {
    pub part_id: u32,
    pub sku: String,
    pub quantity: u32,
    pub unit_price: Money,
}

/// Append-only lifecycle entry.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct StatusChange {
    pub from: RepairStatus,
    pub to: RepairStatus,
    pub changed_by: String,
}

/// A repair job: device in, invoice out.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RepairOrder {
    pub id: u32,
    pub customer_id: u32,
    pub device_id: u32,
    pub status: RepairStatus,
    pub priority: Priority,
    pub labor_minutes: u32,
    pub parts: Vec<OrderPart>,
    pub log: Vec<StatusChange>,
    pub reference_number: u32,
}

impl RepairOrder {
    pub fn new(id: u32, customer_id: u32, device_id: u32) -> Self {
        Self {
            id,
            customer_id,
            device_id,
            status: RepairStatus::Received,
            priority: Priority::Standard,
            labor_minutes: 0,
            parts: Vec::new(),
            log: Vec::new(),
            reference_number: id,
        }
    }

    pub fn with_priority(mut self, priority: Priority) -> Self {
        self.priority = priority;
        self
    }

    pub fn with_labor(mut self, minutes: u32) -> Self {
        self.labor_minutes = minutes;
        self
    }

    /// Move to `next` when the lifecycle allows it, logging the change.
    pub fn transition_to(&mut self, next: RepairStatus, changed_by: &str) -> bool {
        if !self.status.transitions_to().contains(&next) {
            return false;
        }

        self.log.push(StatusChange {
            from: self.status,
            to: next,
            changed_by: changed_by.to_string(),
        });
        self.status = next;

        true
    }

    /// Drive the order to `Completed`, refusing an illegal jump.
    pub fn complete(&mut self, changed_by: &str) -> Result<(), AtelierError> {
        if !self.transition_to(RepairStatus::Completed, changed_by) {
            return Err(AtelierError::IllegalTransition {
                from: self.status,
                to: RepairStatus::Completed,
            });
        }

        Ok(())
    }

    /// Parts-only subtotal; labour is the calculator's business.
    pub fn parts_subtotal(&self) -> Money {
        self.parts
            .iter()
            .map(|line| line.unit_price * i64::from(line.quantity))
            .sum()
    }

    /// Total through the bound strategy: the concrete calculator is decided
    /// by the container, never by this call site.
    pub fn total(&self, container: &Container) -> Money {
        container.invoice_calculator().calculate(self)
    }

    /// Attach a part line, priced from the stock record.
    pub fn add_part(&mut self, part: &Part, quantity: u32) {
        self.parts.push(OrderPart {
            part_id: part.id,
            sku: part.sku.clone(),
            quantity,
            unit_price: part.unit_price,
        });
    }

    /// Still occupying bench space?
    pub fn is_open(&self) -> bool {
        self.status.is_open()
    }
}

impl HasReference for RepairOrder {
    fn reference_number(&self) -> u32 {
        self.reference_number
    }
}
